### ISSUE TITLE
Facebook Presto

### ISSUE THREAD

**Issue by ChrisFeldmeier** [scylladb/scylla#1139]
Is it possible to connect Facebook's prestodb.io to scylla? Has anyone tested this yet?

**Comment by avikivity** [comment]
It's not clear if the connector requires thrift, or whether thrift is optional.  If thrift is optional, it prestodb should work.

**Comment by ChrisFeldmeier** [comment]
I don't know. Maybe the docu does say nothing about that. I have created an issue and waiting for an answer: https://github.com/prestodb/presto/issues/4894

**Comment by dorlaor** [comment]
@tzach can you please post your latest Presto experience here?

**Comment by tzach** [comment]
Presto use CQL, and from my preliminary tests just works with Scylla.
I used the Cassandra connector, and followed the instructions 
https://prestodb.io/docs/current/connector/cassandra.html 

Here some basic commands I tested with 
```
presto:default> SELECT * FROM cassandra.mykeyspace.users where user_id >= 2 and user_id <= 3;
 user_id | fname  | lname 
---------+--------+-------
       2 | dor    | laor  
       3 | shlomi | laor  
(2 rows)

presto:default> SELECT * FROM cassandra.mykeyspace.users where regexp_like(user_id, 'd.*')

presto:default> SELECT * FROM cassandra.mykeyspace.users where regexp_like(fname, 'd.*');
 user_id | fname | lname 
---------+-------+-------
       2 | dor   | laor  
(1 row)

presto:default> select * from (select * from cassandra.mykeyspace.users where lname='laor') where fname='shlomi';
 user_id | fname  | lname 
---------+--------+-------
       3 | shlomi | laor  
(1 row)

presto:default> select lname, count(*) from cassandra.mykeyspace.users group by lname;
  lname   | _col1 
----------+-------
 livyatan |     1 
 laor     |     2 
 livne    |     1 
 kivity   |     1 
(4 rows)
```

Dats used for the above, provisioned directly to Scylla

```
cqlsh> CREATE KEYSPACE mykeyspace WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 1 };
cqlsh> CREATE TABLE users (user_id int PRIMARY KEY, fname text, lname text);
cqlsh:mykeyspace> insert into users (user_id , fname, lname) values (1, 'tzach', 'livyatan');
cqlsh:mykeyspace> insert into users (user_id , fname, lname) values (2, 'dor', 'laor');                                                               
cqlsh:mykeyspace> insert into users (user_id , fname, lname) values (3, 'shlomi', 'laor');                                                            
cqlsh:mykeyspace> insert into users (user_id , fname, lname) values (4, 'shlomi', 'livne');                                                           
cqlsh:mykeyspace> insert into users (user_id , fname, lname) values (6, 'avi', 'kivity'); 
```

**Comment by tzach** [comment]
@ChrisFeldmeier did you managed to run Presto with Scylla?

**Comment by avikivity** [comment]
Closing, as @tzach showed presto+scylla works.

**Comment by ChrisFeldmeier** [comment]
not testet yet

**Comment by gslin** [comment]
I build a cluster today and find Presto 0.149 will try to use Thrift:

    ubuntu@ip-172-31-21-172:~/presto-server-0.149$ ../presto                                                      
    presto> USE cassandra.kktest;
    presto:kktest> SELECT * FROM api_login LIMIT 1;
    Query 20160627_020144_00005_decmp failed: java.io.IOException: Unable to connect to server 172.31.15.98:9160
    presto:kktest> 
    
As expect, port 9042 (Native) is open, and port 9160 (Thrift) is not:

    ubuntu@ip-172-31-21-172:~/presto-server-0.149$ telnet 172.31.15.98 9042
    Trying 172.31.15.98...
    Connected to 172.31.15.98.
    Escape character is '^]'.
    ^]
    telnet> q
    Connection closed.
    ubuntu@ip-172-31-21-172:~/presto-server-0.149$ telnet 172.31.15.98 9160
    Trying 172.31.15.98...
    telnet: Unable to connect to remote host: Connection refused

May I ask @tzach about the settings?

**Comment by tzach** [comment]
Hi @gslin thanks for testing

I restarted with presto 1.4.9 and Scylla 1.2 and its working for me.

Here is a gits with my trivial set up (one presto node)
https://gist.github.com/tzach/31e44b23926e92e3ff3d28630c2ba422

And here is a gits with some test I did, entering data via cqlsh, and query via presto
https://gist.github.com/tzach/7d3a4540264418fdb15aa9fa159e0188

I'm not sure why presto even try to access Thrift in your setup.
Maybe the different is in etc/catalog/cassandra.properties ?

**Comment by ustczen** [comment]
Hi @tzach many thanks for your set up tutorial. I can get a proper output in your sample test.
However, when I test presto + scylla with tons of data, I met the same problem as @gslin ,and presto throw these exception:
`2016-07-09T23:41:36.532+0800	ERROR	Query-20160709_154134_00001_3376r-164	com.facebook.presto.cassandra.CassandraThriftConnectionFactory	Unable to connect to server 127.0.0.1:9160
java.io.IOException: Unable to connect to server 127.0.0.1:9160
	at com.facebook.presto.cassandra.CassandraThriftConnectionFactory.createConnection(CassandraThriftConnectionFactory.java:88)
	at com.facebook.presto.cassandra.CassandraThriftConnectionFactory.getClientFromAddressList(CassandraThriftConnectionFactory.java:68)
	at com.facebook.presto.cassandra.CassandraThriftConnectionFactory.create(CassandraThriftConnectionFactory.java:53)
	at com.facebook.presto.cassandra.CassandraThriftClient.getRangeMap(CassandraThriftClient.java:38)
	at com.facebook.presto.cassandra.CassandraTokenSplitManager.getSplits(CassandraTokenSplitManager.java:65)
	at com.facebook.presto.cassandra.CassandraSplitManager.getSplitsByTokenRange(CassandraSplitManager.java:99)
	at com.facebook.presto.cassandra.CassandraSplitManager.getSplits(CassandraSplitManager.java:82)
	at com.facebook.presto.split.SplitManager.getSplits(SplitManager.java:45)
	at com.facebook.presto.sql.planner.DistributedExecutionPlanner$Visitor.visitTableScan(DistributedExecutionPlanner.java:112)
	at com.facebook.presto.sql.planner.DistributedExecutionPlanner$Visitor.visitTableScan(DistributedExecutionPlanner.java:92)
	at com.facebook.presto.sql.planner.plan.TableScanNode.accept(TableScanNode.java:135)
	at com.facebook.presto.sql.planner.DistributedExecutionPlanner$Visitor.visitFilter(DistributedExecutionPlanner.java:162)
	at com.facebook.presto.sql.planner.DistributedExecutionPlanner$Visitor.visitFilter(DistributedExecutionPlanner.java:92)
	at com.facebook.presto.sql.planner.plan.FilterNode.accept(FilterNode.java:71)
	at com.facebook.presto.sql.planner.DistributedExecutionPlanner$Visitor.visitLimit(DistributedExecutionPlanner.java:258)
	at com.facebook.presto.sql.planner.DistributedExecutionPlanner$Visitor.visitLimit(DistributedExecutionPlanner.java:92)
	at com.facebook.presto.sql.planner.plan.LimitNode.accept(LimitNode.java:86)
	at com.facebook.presto.sql.planner.DistributedExecutionPlanner.plan(DistributedExecutionPlanner.java:78)
	at com.facebook.presto.sql.planner.DistributedExecutionPlanner.plan(DistributedExecutionPlanner.java:83)
	at com.facebook.presto.execution.SqlQueryExecution.planDistribution(SqlQueryExecution.java:303)
	at com.facebook.presto.execution.SqlQueryExecution.start(SqlQueryExecution.java:226)
	at com.facebook.presto.execution.QueuedExecution.lambda$start$1(QueuedExecution.java:62)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.thrift.transport.TTransportException: java.net.ConnectException: Connection refused
	at org.apache.thrift.transport.TSocket.open(TSocket.java:185)
	at org.apache.thrift.transport.TFramedTransport.open(TFramedTransport.java:81)
	at org.apache.cassandra.thrift.TFramedTransportFactory.openTransport(TFramedTransportFactory.java:41)
	at com.facebook.presto.cassandra.CassandraThriftConnectionFactory.createConnection(CassandraThriftConnectionFactory.java:84)
	... 24 more
Caused by: java.net.ConnectException: Connection refused
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)`

**Comment by tzach** [comment]
Hi @ustczen  Thanks for reproducing this issue.
If I'm reading the stack trace correctly, presto try to drop to Thrift, rather than use CQL. Scylla does not support Thrift yet, so the client fails. Scylla will support Thrift very soon (upcoming 1.3 release), but as long as Presto use CQL, it should work now.

Can you please elaborate on your setup, and how much data you loaded before encountering this issue? Also, can you paste your `catalog/cassandra.properties` file?

Thanks

**Comment by gslin** [comment]
I can reproduce in this way:

* Amazon EC2, us-west-2 (Oregan), r3.large with Scylla 1.2.1 AMI (ami-ed4d8b8d).
* Download Presto 0.150 (both server and cli) in this two pages:
 * https://prestodb.io/docs/current/installation/deployment.html
 * https://prestodb.io/docs/current/installation/cli.html
* Use exact https://gist.github.com/tzach/31e44b23926e92e3ff3d28630c2ba422 setting from @tzach's comment.
* Combined logs to cqlsh queries: https://gist.github.com/gslin/c770da45b5c31a1585dd305d3eabf899
* Dataset: https://gist.github.com/gslin/641a6786beb6ecf59d7f5b50e3ae43f1

Create keyspace and table:

    CREATE KEYSPACE mykeyspace WITH replication = {'class':'SimpleStrategy', 'replication_factor':1};
    CREATE TABLE access_log (id UUID PRIMARY KEY, ip TEXT, ident TEXT, user TEXT, timestamp INT, method TEXT, url TEXT, status INT, byte BIGINT, referer TEXT, useragent TEXT);

Use the following command to import:

    cat access.log | ./combined-to-sql.sh | cqlsh

Then run presto-cli with:

    SELECT COUNT(*) FROM cassandra.mykeyspace.access_log;

I can get Thrift error message.

**Comment by tzach** [comment]
@gslin thanks
I was able to reproduce your use case. Look like Presto use both CQL and Thrift for some reason, and choose one base on the case. I will try to dig into Presto to understand if it can be update to use CQL only.

@duarten can you use this case to test the new Scylla Thrift feature?

**Comment by dorlaor** [comment]
@tzach and @gslin, you can try to use the latest nightly binaries which
contain the thrift code

**Comment by ustczen** [comment]
@tzach ,my etc/catalog/cassandra.properties follows your post exactly.
I imported about 100million lines into scylla before I met the exception above.I guess presto use thrift under special circumstances.
I would be appreciate it if you could estimate the release of Scylla 1.3.

**Comment by tzach** [comment]
@ustczen thanks for reporting this issue.
Scylla 1.3 is planned for later this month.

**Comment by dorlaor** [comment]
@ustczen <https://github.com/ustczen> , don't wait to 1.3, test it using
our nightly build.
This way in case something isn't in place we'll fix it for 1.3

**Comment by duarten** [comment]
@tzach Sounds like a good candidate, I'll give it a spin asap.

**Comment by tzach** [comment]
@ustczen @gslin thanks for pointing us to this issue. 
@duarten found that Presto uses Thrift describe_splits verb which was not part of the initial Scylla Thrift implementation. This means Scylla Presto support is limited, and we will fix it following 1.3 release. See #1445 for more.

**Comment by cawallin** [comment]
Hi, I work on Presto -- the Cassandra connector for Presto uses CQL if you are selecting less than cassandra.limit-for-partition-key-select partition keys (by default 200), but if you are selecting more than that, it uses Thrift to figure out how data is partitioned on Cassandra.

**Comment by penberg** [comment]
I think @duarten is working on it.

**Comment by duarten** [comment]
@cawallin Thanks for the explanation!

@penberg Yep, df346d8b4920dcc28fa978ffd94dd8de8271b78d should fix this.

I still have to make sure it's working well with Presto though.

**Comment by penberg** [comment]
@duarten Can you please double-check that all the relevant commits are backported to 1.3? I think they are but it's very easy to make mistakes with that.

**Comment by duarten** [comment]
@penberg They are, yes.

**Comment by tzach** [comment]
@duarten I repeat the test defined by @gslin above and got strange results:

From cqlsh:

```
cqlsh:mykeyspace> select count(*) from access_log ;

 count
-------
  1092

(1 rows)

```

from prest-cli
```
presto:default> SELECT COUNT(*) FROM cassandra.mykeyspace.access_log;
 _col0 
-------
   478 
(1 row)

Query 20160810_132427_00003_z5e3x, FINISHED, 1 node
Splits: 125 total, 125 done (100.00%)
0:02 [478 rows, 478B] [265 rows/s, 265B/s]
```

Same count operation, different results. Can you please take a look

**Comment by duarten** [comment]
@tzach Which Scylla version did you use?

**Comment by tzach** [comment]
> Which Scylla version did you use?

scylla-server-1.3.rc2-20160808

AMI ami-09a63a1e on us-east-1

**Comment by duarten** [comment]
Oh, that's recent enough. I'll take a look asap.

**Comment by duarten** [comment]
@tzach I''m trying to repro, but it just gets stuck here:

```
~/presto-server-0.151 $ ./presto --server localhost:8080 --catalog hive --schema default
presto:default> SELECT COUNT(*) FROM cassandra.mykeyspace.access_log;

Query 20160810_184243_00007_s84mg, PLANNING, 0 nodes, 0 splits
```

Any idea what I'm doing wrong?

**Comment by duarten** [comment]
@tzach Works for me:

```
 ~/presto-server-0.151 $ ./presto --server localhost:8080 --catalog hive --schema default
presto:default> SELECT COUNT(*) FROM cassandra.mykeyspace.access_log;
 _col0
-------
  1092
(1 row)

Query 20160810_184243_00007_s84mg, FINISHED, 1 node
Splits: 258 total, 258 done (100.00%)
1:19 [1.09K rows, 1.07KB] [13 rows/s, 13B/s]
```

I think it's because of https://github.com/scylladb/scylla/commit/325f917d8a86445978f870e294bd98b4dbfa1e5b.

**Comment by tzach** [comment]
@duarten restested with scylla-server-1.3.rc3-20160815.b2e6a52.el7.centos.x86_64

* presto
```
presto:default> SELECT COUNT(*) FROM cassandra.mykeyspace.access_log;
 _col0 
-------
  1085 
(1 row)
```

* cqlsh

```
cqlsh:mykeyspace> select count (*) from access_log;                                                                                                                                      

 count
-------
  1092

(1 rows)
```

**Comment by duarten** [comment]
@tzach I think I fixed the issue (it reliably works for me). Please retest with a build that includes commit 440c1b2 when possible.

**Comment by tzach** [comment]
Thanks @duarten I will test with 1.3 AMI once it is ready

**Comment by penberg** [comment]
@tzach That commit is not in 1.3...

**Comment by avikivity** [comment]
@duarten how does that commit explain the problem?

**Comment by duarten** [comment]
@avikivity WIthout it the client just seems to give up and now report the whole number of rows.

**Comment by avikivity** [comment]
Ah, so it's just performance?

Maybe we should update in the background, so we already return immediately.

**Comment by duarten** [comment]
Right. We update the size estimates in the background, every 5 minutes,
it's just that for some use cases, mostly tests, users might want to see
the stats immediately.

**Comment by tzach** [comment]
@duarten does it mean the presto count will give the right result after 5 min?

**Comment by tzach** [comment]
Created a Docker image for quick test of Scylla+Presto
https://hub.docker.com/r/tzachl/scylla-and-presto-image/

**Comment by duarten** [comment]
@tzach Sorry, had forgotten about your question. It wouldn't work, as the thrift verb was always causing the size estimates to be recorded, on the assumption `describe_splits_ex` wouldn't be called often (Titan calls it with a big range, while Presto calls it often with small ranges), and that just slowed things down enough to cause the Presto client to give up. The fix was to query first, and only trigger a recording of the estimates if the query returns an empty result set. On a related note, I opened #1616.

**Comment by slivne** [comment]
@duarten is this resolved - if so can you reference the commit in head / associated issue that has this info

**Comment by duarten** [comment]
The last commit I did as part of this issue was 9ec939f. For me Presto integration has been working reliably since it.

**Comment by penberg** [comment]
And that commit is in the upcoming 1.3.1 release, so I think we can just close this issue.

**Comment by slivne** [comment]
music to my ears ... closing
